src/t_hash.c

```diff
@@ -621,25 +544,32 @@ robj *hashTypeDup(robj *o) {
 
 /* callback for to check the ziplist doesn't have duplicate records */
 static int _hashZiplistEntryValidation(unsigned char *p, void *userdata) {
+    unsigned char *str;
+    unsigned int slen;
+    long long vll;
+
     struct {
         long count;
         dict *fields;
+        unsigned char **lp;
     } *data = userdata;
 
+    if (!ziplistGet(p, &str, &slen, &vll))
+        return 0;
+
+    sds field = str? sdsnewlen(str, slen): sdsfromlonglong(vll);
+    *(data->lp) = lpAppend(*(data->lp), (unsigned char*)field, sdslen(field));
```

Comment:
for values (when `data->count` is even), it is wasteful to create the sds (at least in the case of strings).
i.e. these sds strings are only needed for the dup search dict on field names.

we can either make some attempt to do this sds creation only for numeric values, and avoid it for string values, or better yet, create an lpAppend variant that takes a long long, and avoid creating sds strings for the values altogether.
i know we discussed this in the past in the context of numerics. but here we also do excessive work on strings.

Reply:
I've added support for inserting integer directly into listpack.